torch/distributed/_tensor/redistribute.py

```diff
@@ -146,18 +202,24 @@ def redistribute_local_tensor(
                 )
                 new_local_tensor = shards[my_coordinate[i]].clone()
             else:
-                # NOTE: this case shouldn't hit _decompose_sharding, decompose sharding should
-                # decompose Shard(0) -> Shard(1) into Shard(0) -> Replicate -> Shard(1)
+                # NOTE: we don't support this case efficiently yet, the fallback path we are going here is
+                # to decompose Shard(0) -> Shard(1) into Shard(0) -> Replicate -> Shard(1)
+                # TODO: enable this with all_to_all
```

Comment:
The 1-d mesh case seems straightforward and we had an implementation. The tricky case is 2-d and more. But I think we can start with the simple 1-d case first. WDYT?

Reply:
Hmmm I think we should not try to enable the easier case and then complicates too much on the redistribute logic, as I couldn't see a clean way to ONLY enable 1D mesh case and throw error for 2D or more. I want to avoid the trap that we enabled 1d but later we could not cleanly handle the multi mesh case. Let's try to enable the alltoall cleanly :) This refactor is trying to make redistribute simpler and could possibly enable for a multi mesh alltoall later.

In addition, the implementation we had probably aren't easily usable anymore as DTensor evolves to use functional collectives, we need to rewrote that with functional collective instead.